flink-connectors/flink-connector-kafka-0.11/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer011ITCase.java

```diff
@@ -566,6 +568,76 @@ public void testRunOutOfProducersInThePool() throws Exception {
 		deleteTestTopic(topic);
 	}
 
+	@Test
+	public void testMigrateFromAtLeastOnceToExactlyOnce() throws Exception {
+		String topic = "testMigrateFromAtLeastOnceToExactlyOnce";
+
+		OperatorSubtaskState producerSnapshot;
+		try (OneInputStreamOperatorTestHarness<Integer, Object> testHarness = createTestHarness(topic, AT_LEAST_ONCE)) {
+			testHarness.setup();
+			testHarness.open();
+			testHarness.processElement(42, 0);
+			testHarness.snapshot(0, 1);
+			testHarness.processElement(43, 2);
+			testHarness.notifyOfCompletedCheckpoint(0);
+			producerSnapshot = testHarness.snapshot(1, 3);
+			testHarness.processElement(44, 4);
+		}
+
+		try (OneInputStreamOperatorTestHarness<Integer, Object> testHarness = createTestHarness(topic, EXACTLY_ONCE)) {
+			testHarness.setup();
+			// restore from snapshot, all records until here should be persisted
+			testHarness.initializeState(producerSnapshot);
+			testHarness.open();
+
+			// write and commit more records
+			testHarness.processElement(44, 7);
+			testHarness.snapshot(2, 8);
```

Comment:
also extend the second try block a little bit by completing at least one checkpoint to make sure that committing also works.